lightning/src/routing/router.rs

```diff
@@ -1045,21 +1075,18 @@ where L::Target: Logger {
 					for chan_id in $node.channels.iter() {
 						let chan = network_channels.get(chan_id).unwrap();
 						if !chan.features.requires_unknown_bits() {
-							if chan.node_one == $node_id {
-								// ie $node is one, ie next hop in A* is two, via the two_to_one channel
-								if first_hops.is_none() || chan.node_two != our_node_id {
-									if let Some(two_to_one) = chan.two_to_one.as_ref() {
-										if two_to_one.enabled {
-											add_entry!(chan_id, chan.node_two, chan.node_one, two_to_one, chan.capacity_sats, &chan.features, $fee_to_target_msat, $next_hops_value_contribution, $next_hops_path_htlc_minimum_msat, $next_hops_path_penalty_msat, $next_hops_cltv_delta);
-										}
-									}
-								}
-							} else {
-								if first_hops.is_none() || chan.node_one != our_node_id{
-									if let Some(one_to_two) = chan.one_to_two.as_ref() {
-										if one_to_two.enabled {
-											add_entry!(chan_id, chan.node_one, chan.node_two, one_to_two, chan.capacity_sats, &chan.features, $fee_to_target_msat, $next_hops_value_contribution, $next_hops_path_htlc_minimum_msat, $next_hops_path_penalty_msat, $next_hops_cltv_delta);
-										}
+							let directed_channel =
+								chan.as_directed_to(&$node_id).expect("inconsistent NetworkGraph");
+							let source = directed_channel.source();
+							let target = directed_channel.target();
```

Comment:
I believe `target` is held on the stack one way or another, so we always have it independent of `DirectedChannelInfo` when we call `channel_penalty_msat`. I believe the same is basically true for `source`, though with a bit more work.

Reply:
Yeah, I don't doubt there is some way to clean this up to remove some excess data while keeping the code readable and supporting passing `DirectedChannelInfo` to `channel_penalty_msat`, which I think is desirable. But perhaps this could be left to a follow-up? Seems the target here is largely coming from `RouteGraphNode`, which itself is holding a copy of the pubkey. I'd imagine we'd want to have `RouteGraphNode` hold a `CandidateRouteHop` instead like `PathBuildingHop` does and have an interface for accessing the source and target for any of the variants. Independently, `DirectedChannelInfo` could use a bool indicating direction instead of two reference to select either, if you prefer to save space over the cost of a condition.

Reply:
> I guess I'm confused, we already pass a `source` and `target` `NodeId` to `channel_penalty_msat`, given that, I don't understand why we care about having a `DirectedChannelInfo` _also_ communicate the source and target `NodeId` - its already there.

My hope was to clean up `channel_penalty_msat`  by passing a `DirectedChannelInfo` instead of `source`, `target`, `capacity_msat`, and later `htlc_maximum_msat`.

> 
> No, I'm mostly looking at `PathBuildingHop` which already holds a `CandidateRouteHop` and trying to squeeze the size of it down as much as possible, and 16 bytes which are currently unused seems like a great way to squeeze it down a bit more.
> 

Won't leaving those 16 bytes in there allow us to remove the 33-byte `node_id` by referencing one in the `CandidateRouteHop`?

Reply:
> My hope was to clean up channel_penalty_msat by passing a DirectedChannelInfo instead of source, target, capacity_msat, and later htlc_maximum_msat.

Right, htlc_maximum_msat needs some additional data from inside the router (if we're already used the hop on a different path), but, sure, I see your point. Given we haven't done it yet, should we just remove the two fields for now, remove the node_id and then revisit once we have a baseline to benchmark against? We can use that baseline as we make the cleanups we want in the future to make sure they dont regress.

> Won't leaving those 16 bytes in there allow us to remove the 33-byte node_id by referencing one in the CandidateRouteHop?

I don't think we have to do something super optimized for it - the `node_id` field is only used as breadcrumbs to walk back the path after we find a path to the sender, so we can do something a bit slow involving branches on the `CandidateRouteHop`, we don't need to store an explicit pointer.

Reply:
Removed the fields as discussed. Will leave the interface cleanup to a follow-up.